### PR TITLE
Auto approve kubelet server certificate signing requests.

### PR DESCRIPTION
**Pull request by jcbsmpsn**
Fixes https://github.com/kubernetes/kubernetes/issues/47208

**Release note**:
```release-note
Adds an approval work flow to the the certificate approver that will approve certificate signing requests from kubelets that meet all the criteria of kubelet server certificates.
```

### CODE REVIEW

**Comment by jcbsmpsn** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by jcbsmpsn** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

File: pkg/controller/certificates/approver/sarapprove.go
Comment (liggitt) [inline pkg/controller/certificates/approver/sarapprove.go]: This doesn't check it is from a node

Comment (liggitt) [inline pkg/controller/certificates/approver/sarapprove.go]: It's missing a check like https://github.com/kubernetes/kubernetes/pull/46884/files#diff-05476437bf221fbe4cab028bf7ffbb74R172

Comment (jcbsmpsn) [inline pkg/controller/certificates/approver/sarapprove.go]: Added.

**Comment by jcbsmpsn** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by jcbsmpsn** [comment]
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this

**Comment by jcbsmpsn** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

File: pkg/controller/certificates/approver/sarapprove.go
Comment (mikedanese) [inline pkg/controller/certificates/approver/sarapprove.go]: Feature flag this so it returns false if the kubelet server cert rotation feature is off.

**Comment by mikedanese** [comment]
You need this:
```diff
diff --git a/cluster/addons/rbac/kubelet-certificate-management.yaml b/cluster/addons/rbac/kubelet-certificate-management.yaml
index 4b4b3d7381..83f1187d57 100644
--- a/cluster/addons/rbac/kubelet-certificate-management.yaml
+++ b/cluster/addons/rbac/kubelet-certificate-management.yaml
@@ -57,5 +57,6 @@ rules:
   - "certificates.k8s.io"
   resources:
   - certificatesigningrequests/selfnodeclient
+  - certificatesigningrequests/selfnodeserver
   verbs:
   - "create"
```

File: pkg/controller/certificates/approver/sarapprove.go
Comment (jcbsmpsn) [inline pkg/controller/certificates/approver/sarapprove.go]: Done.

File: pkg/controller/certificates/approver/sarapprove.go
Comment (mikedanese) [inline pkg/controller/certificates/approver/sarapprove.go]: Invert this. If feature not enabled return false.

File: pkg/kubelet/kubelet.go
Comment (liggitt) [inline pkg/kubelet/kubelet.go]: what key usage or ext key usage does this correspond to?

Comment (mikedanese) [inline pkg/kubelet/kubelet.go]: https://github.com/cloudflare/cfssl/blob/437c6213e569dde639fd116272c27557fd38eaf9/config/config.go#L553

Comment (liggitt) [inline pkg/kubelet/kubelet.go]: oh, digital signature...

```
	"signing":             x509.KeyUsageDigitalSignature,
	"digital signature":   x509.KeyUsageDigitalSignature,
```

Comment (liggitt) [inline pkg/kubelet/kubelet.go]: can we actually change the kubelet to use the usage that corresponds to the RFC? "signing" is pretty ambiguous

File: pkg/controller/certificates/approver/sarapprove.go
Comment (liggitt) [inline pkg/controller/certificates/approver/sarapprove.go]: I'd prefer this to keep the usage that matches the term used in the RFC

**Comment by jcbsmpsn** [comment]
@k8s-bot pull-kubernetes-verify test this

File: pkg/kubelet/kubelet.go
Comment (jcbsmpsn) [inline pkg/kubelet/kubelet.go]: Changed to UsageDigitalSignature is both locations.

File: pkg/controller/certificates/approver/sarapprove.go
Comment (jcbsmpsn) [inline pkg/controller/certificates/approver/sarapprove.go]: Changed back to UsageDigitalSignature.

**Comment by jcbsmpsn** [comment]
@k8s-bot pull-kubernetes-verify test this

**Review by timstclair — APPROVED** [review]
/lgtm

File: pkg/kubelet/kubelet.go
Comment (liggitt) [inline pkg/kubelet/kubelet.go]: I'd really like to see this change make 1.7. I'd prefer not to have a mix of CSR formats to recognize

Comment (jcbsmpsn) [inline pkg/kubelet/kubelet.go]: Working on it.

**Comment by mikedanese** [comment]
@dchen1107 this fixes the alpha e2e tests and only touches alpha codepaths. Please add 1.7 milestone

**Comment by mikedanese** [comment]
We need this to fix broken e2es.

File: pkg/controller/certificates/approver/sarapprove_test.go
Comment (liggitt) [inline pkg/controller/certificates/approver/sarapprove_test.go]: I wasn't expecting this... the approver is permission based... I only expected the kubelet rotation to be gated by the flag. Also, setting this one way in an init() function means it is always enabled when testing, right?

File: pkg/controller/certificates/approver/sarapprove.go
Comment (liggitt) [inline pkg/controller/certificates/approver/sarapprove.go]: I didn't expect the approver to be flag gated this way, only the kubelet function

**Comment by dchen1107** [comment]
I am fine with kubelet change, which is trivial anyway. But I am not comfortable with the changes in pkg/controller/certificates/approver simply due to the lack of the knowledge from my side. I will leave @liggitt to make the final call.

File: pkg/controller/certificates/approver/sarapprove.go
Comment (liggitt) [inline pkg/controller/certificates/approver/sarapprove.go]: if you want to gate it in the apiserver, gate adding isSelfNodeServerCert in `recognizers()`, and you can unit test isSelfNodeServerCert without messing with feature flag enablement in `init()`

Comment (mikedanese) [inline pkg/controller/certificates/approver/sarapprove.go]: I recommended to @jcbsmpsn that we do this. I really don't want to be died to the behaviour of isSelfNodeServer until we have thought through kubelet's self reporting their SANs. What we currently allow is too weak, and this is an API.

File: pkg/controller/certificates/approver/sarapprove_test.go
Comment (mikedanese) [inline pkg/controller/certificates/approver/sarapprove_test.go]: It is only enabled in this package test, yes.

File: pkg/controller/certificates/approver/sarapprove.go
Comment (liggitt) [inline pkg/controller/certificates/approver/sarapprove.go]: ok, then move the gate to recognizers(), and remove the gate-twiddling from the init()

Comment (mikedanese) [inline pkg/controller/certificates/approver/sarapprove.go]: that sounds good to me.

**Comment by liggitt** [comment]
```
I0615 15:22:27.510] Verifying hack/make-rules/../../hack/verify-bazel.sh
I0615 15:22:33.070] --- /tmp/actual-file-768259978	2017-06-15 15:22:33.058595771 -0700
I0615 15:22:33.070] +++ /tmp/expected-file-304640865	2017-06-15 15:22:33.058595771 -0700
I0615 15:22:33.070] @@ -17,10 +17,8 @@
I0615 15:22:33.071]          "//pkg/apis/authorization/v1beta1:go_default_library",
I0615 15:22:33.071]          "//pkg/apis/certificates/v1beta1:go_default_library",
I0615 15:22:33.071]          "//pkg/client/clientset_generated/clientset/fake:go_default_library",
I0615 15:22:33.071] -        "//pkg/features:go_default_library",
I0615 15:22:33.071]          "//vendor/k8s.io/apimachinery/pkg/runtime:go_default_library",
I0615 15:22:33.072]          "//vendor/k8s.io/apimachinery/pkg/runtime/schema:go_default_library",
I0615 15:22:33.072] -        "//vendor/k8s.io/apiserver/pkg/util/feature:go_default_library",
I0615 15:22:33.072]          "//vendor/k8s.io/client-go/testing:go_default_library",
I0615 15:22:33.072]      ],
I0615 15:22:33.072]  )
W0615 15:22:33.173] DRY-RUN: wrote "pkg/controller/certificates/approver/BUILD"
W0615 15:22:33.728] 
W0615 15:22:33.728] 1 BUILD files not up-to-date.
I0615 15:22:33.828] 
I0615 15:22:33.829] Run ./hack/update-bazel.sh
I0615 15:22:33.829] FAILED   hack/make-rules/../../hack/verify-bazel.sh	6s
```

**Comment by liggitt** [comment]
/lgtm

function is gated and off by default, and this fixes the alpha features CI

**Pull request merged**
